### PR TITLE
OP-1136 | Add diagnosis validation upon Admission and Discharge

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OP-1136.

### CODE REVIEW

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 1]
```diff
@@ -343,6 +350,12 @@ protected void validateAdmission(Admission admission, boolean insert) throws OHS
 		}
 		if (admission.getDiseaseIn() == null) {
 			errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisincannotbeempty.msg")));
+		} else {
+			List<Disease> diseases = diseaseManager.getDiseaseIpdIn().stream().filter(dis -> dis == admission.getDiseaseIn()).collect(Collectors.toList());
+			if ( diseases.size() > 0) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:355]:
```suggestion
			if (diseases.size() > 0) {
```

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 2]
```diff
@@ -404,13 +417,27 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
...
 			}
-
+			List<Disease> diseaseIpdOuts1 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut1()).collect(Collectors.toList());
+			List<Disease> diseaseIpdOuts2 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut2()).collect(Collectors.toList());
+			List<Disease> diseaseIpdOuts3 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut3()).collect(Collectors.toList());
 			if (admission.getDiseaseOut1() == null && admission.getDisDate() != null) {
 				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.pleaseselectatleastfirstdiagnosisout.msg")));
 			} else if (admission.getDiseaseOut1() != null && admission.getDisDate() == null) {
 				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.pleaseinsertadischargedate.msg")));
+				throw new OHDataValidationException(errors);
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:427]: Why throw the exception?  That means subsequent tests are not done.  And if this is want you want to do (that is not find everything in one pass) then why isn't there a `throw new ....` after line 424 above?

File: src/test/java/org/isf/admission/test/Tests.java [hunk 3]
```diff
@@ -1287,6 +1290,48 @@ private Admission buildNewAdmission() throws Exception {
...
 				deliveryType, deliveryResult, true);
 	}
+	
+	@Test
+	public void checkAdmisionWithDiseaseIpdIn() throws Exception  {
+		//Given
+		int id = setupTestAdmission(false, true);
+		Admission admission = admissionBrowserManager.getAdmission(id);
+		String diseaseCode = "20";
+		//when
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/admission/test/Tests.java:1300]:
```suggestion
		// When
```
To better match other test's comments

File: src/test/java/org/isf/admission/test/Tests.java [hunk 4]
```diff
@@ -1287,6 +1290,48 @@ private Admission buildNewAdmission() throws Exception {
...
+		Disease diseaseIn1 = diseaseManager.getDiseaseByCode(diseaseCode);
+		Disease diseaseIn11 = admission.getDiseaseIn();
+		admission.setDiseaseIn(diseaseIn1);
+		// Then
+		assertThatThrownBy(() -> admissionBrowserManager.newAdmission(admission))
+				.isInstanceOf(OHDataValidationException.class);
+		admission.setDiseaseIn(diseaseIn11);
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/admission/test/Tests.java:1307]: what is the purpose of this statement given it is the last one in the test?

File: src/test/java/org/isf/admission/test/Tests.java [hunk 5]
```diff
@@ -1287,6 +1290,48 @@ private Admission buildNewAdmission() throws Exception {
...
+		assertThatThrownBy(() -> admissionBrowserManager.updateAdmission(admission))
+				.isInstanceOf(OHDataValidationException.class)
+				.has(
+						new Condition<Throwable>(
+								(e -> ((OHServiceException) e).getMessages().size() == 1), "Expecting single validation error")
+				);
+		admission.setDiseaseOut1(diseaseOut1);
+		admission.setDiseaseOut2(diseaseOut2);
+		admission.setDiseaseOut3(diseaseOut3);
+	}
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/admission/test/Tests.java:1334]: What is the purpose of these last three statements?

Comment (ArnaudFonzam) [reply src/test/java/org/isf/admission/test/Tests.java:1334]: this is because I've noticed that after capturing the assertion, the initial values are refixed. you can look at testMgrValidateMaternity(), for example.

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java (hunk 2 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:427]: it's because it was like this before I only focused on the cases I added.

Comment (dbmalkovsky) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:427]: And you broke the code.  If an error is found and the `throw` happens like you added no other checks are made and a collection object of `errors` would not be needed.   And there would be no need for the not is empty check of the errors collection at the end of this method.   Please remove the `throws` you have added.

Comment (ArnaudFonzam) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:427]: okay

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java (hunk 1 above)
Comment (mwithi) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:354]: The blank space appears to be still there.

Comment (mwithi) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:354]: anyway, the error should be thrown if the diagnosis is NOT found in the stream, hence when `diseases.isEmpty()`, correct?

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 6]
```diff
@@ -404,13 +415,23 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
...
+			List<Disease> diseaseIpdOuts3 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut3()).collect(Collectors.toList());
 			if (admission.getDiseaseOut1() == null && admission.getDisDate() != null) {
 				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.pleaseselectatleastfirstdiagnosisout.msg")));
 			} else if (admission.getDiseaseOut1() != null && admission.getDisDate() == null) {
 				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.pleaseinsertadischargedate.msg")));
 			}
-
+			if (admission.getDiseaseOut1() != null && diseaseIpdOuts1.size() > 0) {
+				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout1isnotallowed.msg")));
+			}
```
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:428]: Same here, the error should be thrown if the disease is NOT found in the list, hence when `diseaseIpdOutsX.isEmpty()`, correct? and the != null check shouldn't be done before the `stream().filter`?

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 7]
```diff
@@ -404,13 +415,23 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
 									DateTimeFormatter.ISO_LOCAL_DATE.format(invalidEnd))));
 				}
 			}
-
+			List<Disease> diseaseIpdOuts1 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut1()).collect(Collectors.toList());
```
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:418]: Formatting...

```suggestion
			List<Disease> diseaseIpdOuts1 = diseaseManager.getDiseaseIpdOut().stream().filter(dis -> dis == admission.getDiseaseOut1()).collect(Collectors.toList());
```

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java (hunk 6 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:428]: If the error is thrown then all subsequent checks are skipped.   I thought the routine was set to check for everything and only throw the exception with the *list* of errors.

Comment (mwithi) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:428]: Sorry, I meant to say to be added to `errors` list...

Comment (ArnaudFonzam) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:428]: If I have understand you suggestion is to check first if diseaseIpdOutsX.isEmpty() check now if admission.getDiseaseOutX() is != null if yes added the error in error list. now if the !diseaseIpdOutsX.isEmpty() now done the stream.filter after check if admission.getDiseaseOut3() != null && diseaseIpdOuts.isEmpty() if yes added the error in error list. Is it please?

File: src/test/java/org/isf/admission/test/Tests.java (hunk 5 above)
Comment (ArnaudFonzam) [reply src/test/java/org/isf/admission/test/Tests.java:1334]: please what about this can I remove or live it?

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 8]
```diff
@@ -343,6 +349,11 @@ protected void validateAdmission(Admission admission, boolean insert) throws OHS
 		}
 		if (admission.getDiseaseIn() == null) {
 			errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisincannotbeempty.msg")));
+		} else {
+			List<Disease> diseases = diseaseManager.getDiseaseIpdIn().stream().filter(dis -> dis == admission.getDiseaseIn()).collect(Collectors.toList());
+			if (diseases.size() > 0) {
+				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisinisnotallowed.msg")));
+			}
```
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:356]: Sorry, I forgot to add this comment: shouldn't the filter compare obects instead of integer, like this `dis -> dis.equals(admission.getDiseaseIn())` ? 

Then, if the `diseaseIn` in the passed `admission` object IS NOT found in the `diseases = diseaseManager.getDiseaseIpdIn()` list, the `diseases` list object will be empty (not greater than zero) and so the error message would make sense, don't you think?

in short:

```suggestion
		} else {
			List<Disease> diseases = diseaseManager.getDiseaseIpdIn().stream().filter(dis -> dis.equals( admission.getDiseaseIn()).collect(Collectors.toList());
			if (diseases.size().isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisinisnotallowed.msg")));
			}
```

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 9]
```diff
@@ -404,13 +415,23 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
 									DateTimeFormatter.ISO_LOCAL_DATE.format(invalidEnd))));
 				}
 			}
-
+			List<Disease> diseaseIpdOuts1 = diseaseManager.getDiseaseIpdOut().stream().filter(dis -> dis == admission.getDiseaseOut1()).collect(Collectors.toList());
+			List<Disease> diseaseIpdOuts2 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut2()).collect(Collectors.toList());
+			List<Disease> diseaseIpdOuts3 = diseaseManager.getDiseaseIpdOut().stream().filter(dis-> dis == admission.getDiseaseOut3()).collect(Collectors.toList());
```
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:420]:
```suggestion
			List<Disease> diseaseIpdOuts1 = diseaseManager.getDiseaseIpdOut().stream().filter(dis -> dis.equals( admission.getDiseaseOut1())).collect(Collectors.toList());
			List<Disease> diseaseIpdOuts2 = diseaseManager.getDiseaseIpdOut().stream().filter(dis -> dis.equals(admission.getDiseaseOut2())).collect(Collectors.toList());
			List<Disease> diseaseIpdOuts3 = diseaseManager.getDiseaseIpdOut().stream().filter(dis -> dis.equals( admission.getDiseaseOut3())).collect(Collectors.toList());
```

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java (hunk 6 above)
Comment (mwithi) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:428]: Sorry, let me close this comment and propose a suggestion

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java (hunk 9 above)
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:420]:
```suggestion
			List<Disease> allowedDiseaseIpdOut = diseaseManager.getDiseaseIpdOut();
			List<Disease> diseaseIpdOuts1 = allowedDiseaseIpdOut .stream().filter(dis -> dis.equals(admission.getDiseaseOut1())).collect(Collectors.toList());
			List<Disease> diseaseIpdOuts2 = allowedDiseaseIpdOut .stream().filter(dis -> dis.equals(admission.getDiseaseOut2())).collect(Collectors.toList());
			List<Disease> diseaseIpdOuts3 = allowedDiseaseIpdOut .stream().filter(dis -> dis.equals(admission.getDiseaseOut3())).collect(Collectors.toList());
```

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 10]
```diff
@@ -404,13 +415,23 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
...
-
+			if (admission.getDiseaseOut1() != null && diseaseIpdOuts1.size() > 0) {
+				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout1isnotallowed.msg")));
+			}
+			if (admission.getDiseaseOut2() != null && diseaseIpdOuts2.size() > 0) {
+				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout2isnotallowed.msg")));
+			}
+			if (admission.getDiseaseOut3() != null && diseaseIpdOuts3.size() > 0) {
+				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout3isnotallowed.msg")));
+			}
```
Comment (mwithi) [inline src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:434]: With the previous suggestion we have that the filtered `diseaseIpdOutsX` list is empty if the `admission.getDiseaseOutX()` is `null` or not found in the `allowedDiseaseIpdOut` list, so this part changes to:

```suggestion
			if (admission.getDiseaseOut1() != null && diseaseIpdOuts1.isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout1isnotallowed.msg")));
			}
			if (admission.getDiseaseOut2() != null && diseaseIpdOuts2.isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout2isnotallowed.msg")));
			}
			if (admission.getDiseaseOut3() != null && diseaseIpdOuts3.isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout3isnotallowed.msg")));
			}
```

But we can optimize further by checking `admission.getDiseaseOutX() != null` first, and the check the stream object together, for instance:

```
			// get list of allowed disease and single diseases only once
			List<Disease> allowedDiseaseIpdOut = diseaseManager.getDiseaseIpdOut();
			Disease diseaseOut1 = admission.getDiseaseOut1();
			Disease diseaseOut2 = admission.getDiseaseOut2();
			Disease diseaseOut3 = admission.getDiseaseOut3();
			
			// if diseaseOut1  is not null then we check if is within the allowed ones, otherwise error
			if (diseaseOut1 != null && allowedDiseaseIpdOut.stream().filter(dis -> dis.equals(diseaseOut1)).collect(Collectors.toList()).isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout1isnotallowed.msg")));
			}

			// repeat for 2
			if (diseaseOut2 != null && allowedDiseaseIpdOut.stream().filter(dis -> dis.equals(diseaseOut2)).collect(Collectors.toList()).isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout2isnotallowed.msg")));
			}

			// repeat for 3
			if (diseaseOut3 != null && allowedDiseaseIpdOut.stream().filter(dis -> dis.equals(diseaseOut3)).collect(Collectors.toList()).isEmpty()) {
				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.diagnosisout3isnotallowed.msg")));
			}
```

File: src/test/java/org/isf/admission/test/Tests.java (hunk 5 above)
Comment (mwithi) [reply src/test/java/org/isf/admission/test/Tests.java:1334]: @dbmalkovsky sounds something we can tackle into OP-911 or different issue?

**Comment by mwithi** [comment]
It seems some tests are failing @ArnaudFonzam

**Comment by ArnaudFonzam** [comment]
> It seems some tests are failing @ArnaudFonzam

yes I am try to fix them

**Review by dbmalkovsky — APPROVED** [review]
LGTM given my one comment that doesn't need to be addressed really

File: src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java [hunk 11]
```diff
@@ -404,108 +415,143 @@ else if (dateOut.isBefore(ad.getAdmDate()) || dateOut.equals(ad.getAdmDate())) {
...
+					errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.admission.pleaseinsertavalidabortdate.msg")));
 				}
 			}
 		}
 		if (!errors.isEmpty()) {
 			throw new OHDataValidationException(errors);
 		}
 	}
+
+	private boolean checkDuplicatedDiseaseOut(Disease diseaseOut1, Disease diseaseOut2, Disease diseaseOut3) {
```
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/admission/manager/AdmissionBrowserManager.java:552]: Yes, I know, it is just improves the code readability. Thanks!
